_posts/2022-12-16-aave-rebasing.md

```diff
@@ -0,0 +1,332 @@
+given proposal, 60% expressed "Yes", and 5% "Abstain", then the aToken contract
+would cast its vote to the UNI governance system with 35% of its weight as
+Against, 60% of its weight as For, and 5% as Abstain.
+
+aUNI holders would effectively get to vote without having a UNI balance!
+
+While flex voting makes this dream scenario possible, it's not without
+technical hurdles.
+The next section will get into what these are and how we solved them.
+
+### Implementing a Flex Voting aToken
+
+If something like the dream scenario were to exist, it would have to take
+account of the fact that aToken balances are constantly increasing, and doing so
+at an ever-changing rate.
+
+After all, if I could have burned my aTokens and withdrawn a certain weight of
+voting tokens at the time of a proposal, any vote cast on my behalf by the aToken
+should reflect that weight.
+
+Let's look at a concrete example:
+
+* Ben deposits 100 UNI into Aave
+* a year passes and aUNI yields 50% interest (it was a _very_ good year)
+* I deposit 100 UNI
+* Ben and I are the only UNI depositors
+* a UNI governance proposal is issued
+* both Ben and I express our votes on the proposal to the aUNI contract
+* Ben expresses a "For" preference
+* I express an "Against" preference
+
+When aUNI casts its vote to the UNI governance system, what should the vote ratios
+be?
+
+If you were only looking at our deposits you might think we
+should have equal voting weight -- since we both deposited 100 UNI.
+But that's wrong.
+Ben should have _more_ voting weight than me because of all of the interest he
+earned over the last year.
+
+Ben's balance at the time of proposal was 150 aUNI.
+This would have entitled him to have withdrawn (and voted with) 150 UNI at that time.
+I, by contrast, could have only withdrawn and voted with the 100 UNI I had just deposited.
+
+So, it seems what we need to look at are _rebased balances_, not deposits.
+The rebased balances would have led us to the correct conclusion: Ben's voting preference should have 50% more voting weight than mine.
+If the aUNI contract holds 250 UNI and Ben is entitled to 150 of that, Ben
+should be able to determine 60% (150/250) of the votes aUNI casts, and I the
+remaining 40% (100/250).
+
+Thus it seems the dream scenario requires us to be able to
+calculate the rebased aToken balance for _any given user_ at _any given block_.
+
+In case it isn't obvious, this is very challenging.
+
+As already mentioned, aToken interest rates are constantly changing
+based on supply and demand within the protocol.
+If supply goes up, interest goes down.
+If demand goes up, interest goes up.
+
+And each time the interest rate changes, the interest accrued since the last update
+is added to the liquidity index (as shown above).
+This means that the aToken liquidity index is constantly changing too.
+
+Recall that the aToken's rebased balance is just the balance on disk multiplied by the liquidity index.
```

Comment:
storage vs disk again (and right below this also)

Reply:
Thanks, done in ed41538